### PR TITLE
[dhcpmon] Open different socket for dual tor to enable interface filtering

### PR DESCRIPTION
**Pull request by kellyyeh**
#### Why I did it
Raw sockets are used for receiving packets and acl rules are not able to block dhcp packets from being received by dhcpmon sockets. 

#### How I did it
In dual tor scenario, a different raw socket is opened and recvfrom is used to filter out interfaces. If the interface mux status is standby, dhcpmon ignores the packet.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

File: src/dhcpmon/src/dhcp_device.cpp [hunk 1]
```diff
@@ -49,34 +51,42 @@
...
+    { 0x45, 8, 0, 0x00001fff },
+    { 0xb1, 0, 0, 0x0000000e },
+    { 0x48, 0, 0, 0x0000000e },
+    { 0x15, 4, 0, 0x00000043 },
+    { 0x15, 3, 0, 0x00000044 },
+    { 0x48, 0, 0, 0x00000010 },
+    { 0x15, 1, 0, 0x00000043 },
+    { 0x15, 0, 1, 0x00000044 },
+    { 0x6, 0, 0, 0x00040000 },
+    { 0x6, 0, 0, 0x00000000 },
```
Comment (yxieca) [inline src/dhcpmon/src/dhcp_device.cpp:89]: Why did you stop using OP_* macros?

File: src/dhcpmon/src/dhcp_device.cpp [hunk 2]
```diff
@@ -221,6 +231,88 @@ static void read_callback(int fd, short event, void *arg)
...
+{
+    dhcp_device_context_t *context = (dhcp_device_context_t*) arg;
+    ssize_t buffer_sz;
+    struct sockaddr_ll sll;
+    socklen_t slen = sizeof sll;
+
+    bool standby = false;
+
+    while ((event == EV_READ) &&
+           ((buffer_sz = recvfrom(fd, context->buffer, context->snaplen, MSG_DONTWAIT, (struct sockaddr *)&sll, &slen)) > 0)) {
```
Comment (yxieca) [inline src/dhcpmon/src/dhcp_device.cpp:255]: Please put '{' in a new line when while condition extends multiple lines.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 3]
```diff
@@ -221,6 +231,88 @@ static void read_callback(int fd, short event, void *arg)
...
+        std::string member_table = std::string("VLAN_MEMBER|") + context->intf + "|";
+        char interfaceName;
+  	    char *interface = if_indextoname(sll.sll_ifindex, &interfaceName);
```
Comment (yxieca) [inline src/dhcpmon/src/dhcp_device.cpp:258]: indentation issue? Did you have tab here?

File: src/dhcpmon/src/dhcp_device.cpp [hunk 4]
```diff
@@ -221,6 +231,88 @@ static void read_callback(int fd, short event, void *arg)
...
+        mStateDbMuxTablePtr->hget(intf, "state", state);
+        if (state != "standby" && configDb.exists(member_table.append(interface))) {
+            struct ether_header *ethhdr = (struct ether_header*) context->buffer;
+            struct ip *iphdr = (struct ip*) (context->buffer + IP_START_OFFSET);
+            struct udphdr *udp = (struct udphdr*) (context->buffer + UDP_START_OFFSET);
+            uint8_t *dhcphdr = context->buffer + DHCP_START_OFFSET;
+            int dhcp_option_offset = DHCP_START_OFFSET + DHCP_OPTIONS_HEADER_SIZE;
+
+            if ((buffer_sz > UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) &&
+                (ntohs(udp->len) > DHCP_OPTIONS_HEADER_SIZE)) {
```
Comment (yxieca) [inline src/dhcpmon/src/dhcp_device.cpp:271]: '{' in a new line please.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 5]
```diff
@@ -221,6 +231,88 @@ static void read_callback(int fd, short event, void *arg)
...
+                int dhcp_sz = ntohs(udp->len) < buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr) ?
+                            ntohs(~udp->len) : buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr);
```
Comment (yxieca) [inline src/dhcpmon/src/dhcp_device.cpp:273]: What will negative dhcp_sz do?

Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp:275]: If dhcp_sz is somehow negative, while loop below won't be executed

Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp]: Updated dhcp_sz to be 0

File: src/dhcpmon/src/dhcp_device.cpp [hunk 6]
```diff
@@ -185,10 +186,10 @@ static void read_callback(int fd, short event, void *arg)
         uint8_t *dhcphdr = context->buffer + DHCP_START_OFFSET;
         int dhcp_option_offset = DHCP_START_OFFSET + DHCP_OPTIONS_HEADER_SIZE;
 
-        if ((buffer_sz > UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) &&
+        if (((unsigned)buffer_sz > UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) &&
             (ntohs(udp->len) > DHCP_OPTIONS_HEADER_SIZE)) {
             int dhcp_sz = ntohs(udp->len) < buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr) ?
-                          ntohs(udp->len) : buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr);
+                          ntohs(udp->len) : 0;
```
Comment (yxieca) [inline src/dhcpmon/src/dhcp_device.cpp:192]: The original code was ok.

**Review by yxieca — APPROVED** [review]

**Pull request merged**